proxylib/memcached/text/parser.go

```diff
@@ -0,0 +1,329 @@
+	"github.com/cilium/cilium/proxylib/proxylib"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// ParserFactory implements proxylib.ParserFactory
+type ParserFactory struct{}
+
+// Create creates binary memcached parser
+func (p *ParserFactory) Create(connection *proxylib.Connection) proxylib.Parser {
+	log.Infof("ParserFactory: Create: %v", connection)
+	return &Parser{connection: connection, replyQueue: make([]*replyIntent, 0)}
+}
+
+// compile time check for interface implementation
+var _ proxylib.ParserFactory = &ParserFactory{}
+
+// ParserFactoryInstance creates text parser for unified memcached parser
+var ParserFactoryInstance *ParserFactory
+
+// Parser implements proxylib.Parser
+type Parser struct {
+	connection *proxylib.Connection
+
+	requestCount uint32
+	replyCount   uint32
+	replyQueue   []*replyIntent
+
+	//set to true when watch command is observed
+	watching bool
+}
+
+type replyIntent struct {
+	command   []byte
+	requestID uint32
+	denied    bool
+}
+
+var _ proxylib.Parser = &Parser{}
+
+const (
+	CAS_WITH_NOREPLY_FIELDS     = 7
+	STORAGE_WITH_NOREPLY_FIELDS = 6
+	DELETE_WITH_NOREPLY_FIELDS  = 3
+	INCR_WITH_NOREPLY_FIELDS    = 4
+	TOUCH_WITH_NOREPLY_FIELDS   = 4
+)
+
+// OnData parses text memcached data
+func (p *Parser) OnData(reply, endStream bool, dataBuffers [][]byte) (proxylib.OpType, int) {
+	if reply {
+		if p.injectFromQueue() {
+			return proxylib.INJECT, len(DeniedMsg)
+		}
+		if len(dataBuffers) == 0 {
+			return proxylib.NOP, 0
+		}
+	}
+
+	// TODO: don't copy data to new slices
+	data := (bytes.Join(dataBuffers, []byte{}))
+	log.Debugf("Data length: %d", len(data))
+
+	linefeed := bytes.Index(data, []byte("\r\n"))
+	if linefeed < 0 {
+		log.Debugf("Did not receive full first line")
+		if data[len(data)-1] == '\r' {
+			return proxylib.MORE, 1
+		} else {
+			return proxylib.MORE, 2
+		}
+	}
+
+	// TODO: iterate over data without copying it to new slices
+	// Tokenizing in memcached is done by spaces: https://github.com/memcached/memcached/blob/master/memcached.c#L2978
+	tokens := bytes.Fields(data[:linefeed])
+
+	if !reply {
+		meta := meta.MemcacheMeta{
+			Command: string(tokens[0]),
+		}
+		command := tokens[0]
+
+		frameLength := 0
+		noreply := false
+		if p.isCommandRetrieval(command) {
+			// get, gets, gat, gats
+			if bytes.HasPrefix(command, []byte("get")) {
+				meta.Keys = tokens[1:]
+			} else if bytes.HasPrefix(command, []byte("gat")) {
+				meta.Keys = tokens[2:]
+			}
+			frameLength = linefeed + 2
+		} else if p.isCommandStorage(command) {
+			// storage commands
+			meta.Keys = tokens[1:2]
+			nBytes, err := strconv.Atoi(string(tokens[4]))
+			if err != nil {
+				log.Error("Failed to parse storage payload length")
+				return proxylib.ERROR, 0
+			}
+			// 2 additional bytes for terminating linefeed
+			frameLength = linefeed + 2 + nBytes + 2
+
+			if command[0] == 'c' { //storage command is "cas"
+				noreply = len(tokens) == CAS_WITH_NOREPLY_FIELDS
```

Comment:
@nebril the length of a string would help to understand what you are comparing the tokens with. Aren't "tonkens" supposed to be strings? For this reason it would be easier to understand the line
```
       noreply = len(tokens) == len("special_tokenized_string")
```

Reply:
tokens are parts of memcached command separated by spaces. 
`noreply` is an optional part of the command (which is usually added at the end). So if number of tokens in command matches version of command with `noreply`, the parser assumes that client wants noreply version and sets the variable accordingly.

https://github.com/memcached/memcached/blob/master/doc/protocol.txt